ntp-os-clock/src/unix.rs

```diff
@@ -0,0 +1,402 @@
+// Note on unsafe usage.
+//
+// This module uses unsafe code to interact with the system calls that
+// are used to set/get the current behaviour and time of the clock. It
+// is constructed in such a way that use of the public functions is
+// safe regardless of given arguments.
+
+use std::{ffi::CStr, os::unix::prelude::RawFd};
+
+use crate::{Error, EPOCH_OFFSET};
+use ntp_proto::{NtpClock, NtpDuration, NtpLeapIndicator, NtpTimestamp, PollInterval};
+
+// Libc has no good other way of obtaining this, so let's at least make our functions
+// more readable.
+#[cfg(target_os = "linux")]
+pub(crate) const EMPTY_TIMEX: libc::timex = libc::timex {
+    modes: 0,
+    offset: 0,
+    freq: 0,
+    maxerror: 0,
+    esterror: 0,
+    status: 0,
+    constant: 0,
+    precision: 0,
+    tolerance: 0,
+    time: libc::timeval {
+        tv_sec: 0,
+        tv_usec: 0,
+    },
+    tick: 0,
+    ppsfreq: 0,
+    jitter: 0,
+    shift: 0,
+    stabil: 0,
+    jitcnt: 0,
+    calcnt: 0,
+    errcnt: 0,
+    stbcnt: 0,
+    tai: 0,
+    __unused1: 0,
+    __unused2: 0,
+    __unused3: 0,
+    __unused4: 0,
+    __unused5: 0,
+    __unused6: 0,
+    __unused7: 0,
+    __unused8: 0,
+    __unused9: 0,
+    __unused10: 0,
+    __unused11: 0,
+};
+
+#[cfg(any(target_os = "freebsd", target_os = "macos"))]
+pub(crate) const EMPTY_TIMEX: libc::timex = libc::timex {
+    modes: 0,
+    offset: 0,
+    freq: 0,
+    maxerror: 0,
+    esterror: 0,
+    status: 0,
+    constant: 0,
+    precision: 0,
+    tolerance: 0,
+    ppsfreq: 0,
+    jitter: 0,
+    shift: 0,
+    stabil: 0,
+    jitcnt: 0,
+    calcnt: 0,
+    errcnt: 0,
+    stbcnt: 0,
+};
+
+/// NTP Clock that uses the unix NTP KAPI clock functions to get/modify the
+/// current time.
+// Implementation note: this is intentionally a bare struct, the NTP Clock defined
+// in the NTP KAPI is unique and no state is needed to interact with it.
+#[derive(Debug, Default, Clone)]
+pub struct UnixNtpClock {
+    clock: libc::clockid_t,
+}
+
+impl UnixNtpClock {
+    pub fn realtime() -> Self {
+        Self::custom(libc::CLOCK_REALTIME)
+    }
+
+    pub fn custom(id: libc::clockid_t) -> Self {
+        Self { clock: id }
+    }
+
+    #[cfg_attr(target_os = "linux", allow(unused))]
+    pub fn ptp0() -> Result<Self, Error> {
+        let path = CStr::from_bytes_with_nul(b"/dev/ptp0\0").unwrap();
+        // SAFETY
+        //
+        // we know this is a file descriptor of a clock
+        unsafe { Self::from_path(path) }
```

Comment:
How do we know this? a user could have just created a regular file under /dev/ptp0